### PR TITLE
Make Script::fmt_asm a static method and add Script::str_asm 

### PR DESCRIPTION
**Pull request by stevenroose**
This makes it convenient to print/construct the script assembly on byte slices withoout having to clone them to copy them to create a Script struct.

### CODE REVIEW

File: src/blockdata/script.rs
Comment (stevenroose) [reply src/blockdata/script.rs:511]: Hmm yeah naming. I figured not too important is't not really gonna be a high-usage method. `asm` is a member method, that would be used more. I made `str_asm` kinda cfr `fmt_asm`, but `asm_from_bytes` makes sense.

Comment (dr-orlovsky) [reply src/blockdata/script.rs:511]: I think the good naming for the associated conversion function should include both original and return data type, like `bytes_to_string` (`str` is reserved for `&str` return type). Thus, it will be `Script::bytes_to_string`, which is very readable and explicit.

File: src/blockdata/script.rs
Comment (dr-orlovsky) [inline src/blockdata/script.rs:44]: Why not to use this opportunity and derive standard things for Script? We can do `#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]`

File: src/blockdata/script.rs
Comment (dr-orlovsky) [inline src/blockdata/script.rs:442]: Why not `impl AsRef<[u8]>`?

**Review by dr-orlovsky — CHANGES_REQUESTED** [review]
Agree with @sgeisler in his requested changes

File: src/blockdata/script.rs
Comment (sgeisler) [reply src/blockdata/script.rs:511]: I disagree on the return type being in the name, `string` is as non-descriptive as it gets (could be hex for all we know), this should remain the chosen encoding, i.e. `asm`. I'd be ok with `bytes_to_asm`, but don't really care as long as it's descriptive.

File: src/blockdata/script.rs
Comment (sgeisler) [reply src/blockdata/script.rs:44]: I mean we totally could do that here, but it's up to @stevenroose if he wants it in, it's kinda unrelated.

File: src/blockdata/script.rs
Comment (sgeisler) [reply src/blockdata/script.rs:442]: Can we already do `impl` with 1.29?  I wouldn't be too opposed to making things more generic like that, but otoh if something already is `AsRef<[u8]>` you can just call `as_ref()` on it when calling `fmt_asm`, it's just some very marginal convenience.

It's sad the following doesn't work:

```rust
struct Script(Vec<u8>);

impl AsRef<[u8]> for Script {
    fn as_ref(&self) -> &[u8] {
        &self.0
    }
}

impl Script {
    fn asm(self: impl AsRef<[u8]>) -> String {
        unimplemented!();
    }
}
```

This unfortunately only seems to work for types that directly deref to `Self`:
```
   = note: type of `self` must be `Self` or a type that dereferences to it
   = help: consider changing to `self`, `&self`, `&mut self`, `self: Box<Self>`, `self: Rc<Self>`, `self: Arc<Self>`, or `self: Pin<P>` (where P is one of the previous types except `Self`)
```

File: src/blockdata/script.rs
Comment (dr-orlovsky) [reply src/blockdata/script.rs:511]: Agree on `bytes_to_asm`

File: src/blockdata/script.rs
Comment (stevenroose) [reply src/blockdata/script.rs:442]: Is that worth the effort, you think? The fmt asm method is really already a helper that I don't think has any real users. I can wait until we do a breaking release for this to be included.

File: src/blockdata/script.rs
Comment (stevenroose) [reply src/blockdata/script.rs:44]: It probably belongs in another PR. Though I definitely like those proposed changes.

File: src/blockdata/script.rs
Comment (stevenroose) [reply src/blockdata/script.rs:442]: Yeah I agree it it's a marginal benefit for a barely used API. Better keep it more readable.

**Review by sanket1729 — APPROVED** [review]
utACK 2cb81854aa7d89524f9479d7c228a86b7cc0e8b1 . I have no strong opinions on naming, as 1) it is a debug/test method and 2) can be easily seen by looking up docs. 

It will be a breaking change for rust-miniscript because we use this in our tests, but it should not be hard to fix.

File: src/blockdata/script.rs
Comment (sgeisler) [reply src/blockdata/script.rs:442]: As it is this is a breaking change as far as I can tell since `fmt_asm` is public. I think this little change could make this PR non-breaking, allowing us to change the milestone to 0.26.1. Afaik we wanted to get away from frequent API breaks.

File: src/blockdata/script.rs
Comment (sgeisler) [reply src/blockdata/script.rs:511]: @stevenroose Would you be ok with giving this a more descriptive name? Rarely used or not, it is part of the public API we will be stuck with for a while.

File: src/blockdata/script.rs
Comment (stevenroose) [reply src/blockdata/script.rs]: ack

**Comment by stevenroose** [comment]
I changed the names. And rebased.

**Review by apoelstra — APPROVED** [review]
ack 851a3a15c0a124ff9e41e542b5445b4d59786b40

**Review by dr-orlovsky — APPROVED** [review]
Ok, agree with this version

**Review by sgeisler — APPROVED** [review]
tACK 851a3a15c0a124ff9e41e542b5445b4d59786b40

<details>
  <summary>Test log:</summary>

```
Mar 27 11:55:30.957  INFO testinator: Installing rust toolchain 'nightly'
Mar 27 11:55:49.972  INFO testinator: Installing rust toolchain 'stable'
Mar 27 11:56:54.896  INFO testinator: Installing rust toolchain '1.29.0'
Mar 27 11:56:55.916  INFO testinator: Preparing environment for rust nightly tests (9 configurations)
Mar 27 11:56:55.916  INFO testinator: Preparing environment for rust stable tests (8 configurations)
Mar 27 11:56:55.916  INFO testinator: Preparing environment for rust 1.29.0 tests (8 configurations)
Mar 27 11:59:19.513  INFO testinator: Running rust 1.29.0 tests in /tmp/rust-bitcoin-1.29.0.Cf3vnJcUsHEA/rust-bitcoin
Mar 27 11:59:19.513 DEBUG testinator: Generating lock file with rust=1.29.0
Mar 27 11:59:19.519  INFO testinator: Running rust nightly tests in /tmp/rust-bitcoin-nightly.o31ZRDPsAqge/rust-bitcoin
Mar 27 11:59:19.522  INFO testinator: Running rust stable tests in /tmp/rust-bitcoin-stable.e6g5fj0QAChO/rust-bitcoin
Mar 27 12:00:19.701 DEBUG testinator: Pinning cc to 1.0.41
Mar 27 12:00:20.015 DEBUG testinator: Pinning serde to 1.0.98
Mar 27 12:00:20.167 DEBUG testinator: Pinning serde_derive to 1.0.98
Mar 27 12:00:20.319 DEBUG testinator: Pinning byteorder to 1.3.4
Mar 27 12:01:18.231  INFO testinator: Test rust=nightly, features=[secp-recovery] succeeded!
Mar 27 12:01:25.878  INFO testinator: Test rust=stable, features=[secp-recovery] succeeded!
Mar 27 12:01:54.922  INFO testinator: Test rust=1.29.0, features=[secp-recovery] succeeded!
Mar 27 12:02:23.395  INFO testinator: Test rust=nightly, features=[use-serde] succeeded!
Mar 27 12:02:45.540  INFO testinator: Test rust=stable, features=[use-serde] succeeded!
Mar 27 12:02:52.858  INFO testinator: Test rust=nightly, features=[base64] succeeded!
Mar 27 12:03:17.262  INFO testinator: Test rust=nightly, features=[secp-lowmemory] succeeded!
Mar 27 12:03:17.285  INFO testinator: Test rust=stable, features=[base64] succeeded!
Mar 27 12:03:58.190  INFO testinator: Test rust=1.29.0, features=[use-serde] succeeded!
Mar 27 12:03:58.738  INFO testinator: Test rust=nightly, features=[rand] succeeded!
Mar 27 12:03:58.870  INFO testinator: Test rust=stable, features=[secp-lowmemory] succeeded!
Mar 27 12:04:31.496  INFO testinator: Test rust=nightly, features=[unstable] succeeded!
Mar 27 12:04:34.408  INFO testinator: Test rust=stable, features=[rand] succeeded!
Mar 27 12:04:50.245  INFO testinator: Test rust=1.29.0, features=[base64] succeeded!
Mar 27 12:05:01.139  INFO testinator: Test rust=nightly, features=[bitcoinconsensus] succeeded!
Mar 27 12:05:22.010  INFO testinator: Test rust=stable, features=[bitcoinconsensus] succeeded!
Mar 27 12:05:37.926  INFO testinator: Test rust=nightly, features=[secp-recovery,use-serde,base64,secp-lowmemory,rand,unstable,bitcoinconsensus] succeeded!                                                                                                  
Mar 27 12:05:43.496  INFO testinator: Test rust=1.29.0, features=[secp-lowmemory] succeeded!
Mar 27 12:05:59.844  INFO testinator: Test rust=nightly, features=[] succeeded!
Mar 27 12:06:08.202  INFO testinator: Test rust=stable, features=[secp-recovery,use-serde,base64,secp-lowmemory,rand,bitcoinconsensus] succeeded!                                                                                                            
Mar 27 12:06:30.850  INFO testinator: Test rust=1.29.0, features=[rand] succeeded!
Mar 27 12:06:33.799  INFO testinator: Test rust=stable, features=[] succeeded!
Mar 27 12:07:20.275  INFO testinator: Test rust=1.29.0, features=[bitcoinconsensus] succeeded!
Mar 27 12:08:10.412  INFO testinator: Test rust=1.29.0, features=[secp-recovery,use-serde,base64,secp-lowmemory,rand,bitcoinconsensus] succeeded!                                                                                                            
Mar 27 12:08:52.316  INFO testinator: Test rust=1.29.0, features=[] succeeded!
Mar 27 12:08:56.460  INFO testinator: Fuzzing deserialize_script
Mar 27 12:10:10.226  INFO testinator: Successfully fuzzed deserialize_script
Mar 27 12:10:10.226  INFO testinator: Fuzzing uint128_fuzz
Mar 27 12:11:12.186  INFO testinator: Successfully fuzzed uint128_fuzz
Mar 27 12:11:12.186  INFO testinator: Fuzzing deserialize_amount
Mar 27 12:12:13.210  INFO testinator: Successfully fuzzed deserialize_amount
Mar 27 12:12:13.211  INFO testinator: Fuzzing deserialize_transaction
Mar 27 12:13:15.198  INFO testinator: Successfully fuzzed deserialize_transaction
Mar 27 12:13:15.198  INFO testinator: Fuzzing deser_net_msg
Mar 27 12:14:18.252  INFO testinator: Successfully fuzzed deser_net_msg
Mar 27 12:14:18.253  INFO testinator: Fuzzing deserialize_address
Mar 27 12:15:19.242  INFO testinator: Successfully fuzzed deserialize_address
Mar 27 12:15:19.242  INFO testinator: Fuzzing deserialize_block
Mar 27 12:16:21.104  INFO testinator: Successfully fuzzed deserialize_block
Mar 27 12:16:21.104  INFO testinator: Fuzzing outpoint_string
Mar 27 12:17:22.236  INFO testinator: Successfully fuzzed outpoint_string
Mar 27 12:17:22.236  INFO testinator: Fuzzing deserialize_psbt
Mar 27 12:18:25.179  INFO testinator: Successfully fuzzed deserialize_psbt
```
</details>

**Comment by sgeisler** [comment]
I also think this version isn't API breaking anymore, am I right? @stevenroose

**Pull request merged**
